vulnerable_people_form/integrations/persistence.py

```diff
@@ -0,0 +1,199 @@
+import boto3
+
+from flask import current_app
+
+
+def get_rds_data_client(app=current_app):
+    return boto3.client(
+        "rds-data", endpoint_url=app.config.get("LOCAL_AWS_ENDPOINT_URL")
+    )
+
+
+def get_rds_client(app=current_app):
+    return boto3.client("rds", endpoint_url=app.config.get("LOCAL_AWS_ENDPOINT_URL"))
+
+
+def get_secretsmanager_client(app=current_app):
+    return boto3.client(
+        "secretsmanager", endpoint_url=app.config.get("LOCAL_AWS_ENDPOINT_URL")
+    )
+
+
+def _find_database_arn(app):
+    rds = get_rds_client(app=app)
+    clusters = rds.describe_db_clusters()["DBClusters"]
+    coronavirus_rds = [
+        x
+        for x in clusters
+        if x["DBClusterIdentifier"].startswith(app.config["DATABASE_CLUSTER_PREFIX"])
+    ][0]
+    return coronavirus_rds["DBClusterArn"]
+
+
+def _find_database_secret_arn(app):
+    secrets_manager = get_secretsmanager_client(app=app)
+    return secrets_manager.list_secrets(
+        Filters=[{"Key": "tag-value", "Values": app.config["DATABASE_SECRET_TAGS"]}]
+    )["SecretList"][0]["ARN"]
+
+
+def init_app(app):
+    if app.config.get("AWS_RDS_DATABASE_ARN_OVERRIDE"):
+        app.config["AWS_RDS_DATABASE_ARN"] = app.config["AWS_RDS_DATABASE_ARN_OVERRIDE"]
+    else:
+        app.config["AWS_RDS_DATABASE_ARN"] = _find_database_arn(app)
+
+    if app.config.get("AWS_RDS_SECRET_ARN_OVERRIDE"):
+        app.config["AWS_RDS_SECRET_ARN"] = app.config["AWS_RDS_SECRET_ARN_OVERRIDE"]
+    else:
+        app.config["AWS_RDS_SECRET_ARN"] = _find_database_secret_arn(app)
+
+
+def generate_string_parameter(name, value):
+    return {
+        "name": name,
+        "value": {"isNull": True} if value is None else {"stringValue": value,},
+    }
+
+
+def generate_int_parameter(name, value):
+    return {
+        "name": name,
+        "value": {"doubleValue": value,},
+    }
+
+
+def generate_bigint_parameter(name, value):
+    return {
+        "name": name,
+        "value": {"longValue": value,},
+    }
+
+
+def generate_date_parameter(name, value):
+    return {
+        "name": name,
+        "typeHint": "DATE",
+        "value": {"stringValue": "{year}-{month:0>2}-{day:0>2}".format(**value),},
+    }
+
+
+def execute_sql(sql, parameters):
+    return get_rds_data_client().execute_statement(
```

Comment:
The SQL mode needs to be set to strict. e.g
`SET SESSION sql_mode = 'STRICT_ALL_TABLES';`

Reply:
The Aurora HTTP endpoint only accepts single statements. What we could do is open a transaction, get a transaction id, execute the two statements in two separate HTTP requests, then issue a 4th HTTP request to close the transaction. However, this seems rather a lot of effort.

Would it be better to set `sql_mode` at the instance_level instead, rather than relying on clients to implement correct behaviour?

https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Reference.html#AuroraMySQL.Reference.ParameterGroups

Reply:
Yeah, we're working on setting `sql_mode` at the global level. There's a branch with it being worked on over here https://github.com/alphagov/covid-engineering/compare/enable-mysql-strict-mode#diff-a36f48a6afdedf2f324409d542bf5c82

I think at the moment I'd rather just do the four requests until we can get the setting of the global `sql_mode` done, then we can undo this work if it turns out it's too slow.

My preference is that I'd rather it be slower but less likely to allow bad data to be ingested (until we get the global option set).

Reply:
This is now done 🙂